ignitionrobotics/ign-physics
# [Citadel] Add all features to tpeplugin

This PR add all the features to `tpeplugin` and enables testing with example worlds. This PR should be merged after #45 (adds needed functions to `tpelib`) and #32 (implements `Base` and `EntityManagementFeatures`). 

A few things need to be addressed by follow-up PRs

- Add collision detection
- Add a `VelocityCmd` to `ign-gazebo/src/systems/physics/Physics.cc` to enable setting world velocity without applying force to joint 

So far, any models without joint should be able to transform pose with TPE.

To test with ignition gazebo 

```bash
cd src/ign-gazebo
git checkout chapulina/tpe
cd ~/citadel_ws
colcon build --cmake-args -DBUILD_TESTING=OFF --merge-install --packages-up-to ignition-gazebo3
. ~/citadel_ws/install/setup.bash
ign gazebo -v 2 shapes.sdf --physics-engine ignition-physics-tpe-plugin
```
Terminal output
```bash
[GUI] [Wrn] [Application.cc:649] [QT] file::/ComponentInspector/ComponentInspector.qml:161: ReferenceError: ComponentsModel is not defined
construct model ground_plane
construct model box
construct model cylinder
construct model sphere
construct link link
construct link box_link
construct link cylinder_link
construct link sphere_link
construct collision collision
construct collision box_collision
construct collision cylinder_collision
construct collision sphere_collision
```

The shapes no longer collapse to origin after the commit https://github.com/ignitionrobotics/ign-physics/pull/32/commits/344f7f48cd16841a5d683c43a72706fce28a54c4 , and warnings are fixed by commit https://github.com/ignitionrobotics/ign-physics/pull/32/commits/01bf3deacd965724a19d1ede8ca1333f5541f18a 
Shapes can be moved around by using transform tools after commit https://github.com/ignitionrobotics/ign-physics/pull/32/commits/ad391b3b4eb6a0299b3bf59ba9ca76aded48bc69

Demo
![TPE_shapes_demo](https://user-images.githubusercontent.com/22240514/80547524-4af17480-896d-11ea-9181-5d5705e7d0e0.gif)

I also tried with a couple other example worlds. See performance summary below:

- `actor.sdf`: loads fine, actors move, warns missing joint entity, can't transform pose
- `apply_joint_force.sdf`: loads fine, warns missing joint entity, can't transform pose
- `breadcrumbs.sdf`: loads fine, warns missing joint, can't transform pose
- `camera_sensor.sdf`: loads fine, sensor works, cube transforms pose, cone doesn’t and warns missing joint entity
- `contact_sensor`: loads fine, transforms pose, sensor works
- `fuel.sdf`: loads fine, warns missing joint entity, actor can move, can't transform pose
- `tunnel.sdf`: loads fine, warns missing joint entity, car transforms pose
